From a57ea7eda73f5bdf5359d585762275d4f7282b54 Mon Sep 17 00:00:00 2001
From: Aikar <aikar@aikar.co>
Date: Wed, 4 Jul 2018 03:39:51 -0400
Subject: [PATCH] Avoid Chunk Lookups for Entity/TileEntity Current Chunk

In many places where we simply want the current chunk the entity
is in, instead of doing a hashmap lookup for it, we now have access
to the object directly on the Entity/TileEntity object we can directly grab.

Use that local value instead to reduce lookups in many hot places.

diff --git a/src/main/java/net/minecraft/server/Chunk.java b/src/main/java/net/minecraft/server/Chunk.java
index 7a7d656926..b37fa3829b 100644
--- a/src/main/java/net/minecraft/server/Chunk.java
+++ b/src/main/java/net/minecraft/server/Chunk.java
@@ -721,6 +721,7 @@ public class Chunk implements IChunkAccess {
         ((HeightMap) this.heightMap.get(heightmap_type)).a(along);
     }
 
+    public void removeEntity(Entity entity) { b(entity); } // Paper - OBFHELPER
     public void b(Entity entity) {
         this.a(entity, entity.af);
     }
diff --git a/src/main/java/net/minecraft/server/World.java b/src/main/java/net/minecraft/server/World.java
index 00eb342f8c..127dcedc97 100644
--- a/src/main/java/net/minecraft/server/World.java
+++ b/src/main/java/net/minecraft/server/World.java
@@ -1208,12 +1208,15 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
         int j;
         // Paper start - Set based removal lists
         for (Entity e : this.g) {
+            /*
             j = e.getChunkZ();
             int k = e.getChunkX();
 
             if (e.inChunk && this.isChunkLoaded(k, j, true)) {
                 this.getChunkAt(k, j).b(e);
-            }
+            }*/
+            Chunk chunk = e.inChunk ? e.getCurrentChunk() : null;
+            if (chunk != null) chunk.removeEntity(e);
         }
 
         for (Entity e : this.g) {
@@ -1274,12 +1277,17 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
             this.methodProfiler.e();
             this.methodProfiler.a("remove");
             if (entity.dead) {
+                // Paper start
+                /*
                 j = entity.ae;
                 int l = entity.ag;
 
                 if (entity.inChunk && this.isChunkLoaded(j, l, true)) {
                     this.getChunkAt(j, l).b(entity);
-                }
+                }*/
+                Chunk chunk = entity.inChunk ? entity.getCurrentChunk() : null;
+                if (chunk != null) chunk.removeEntity(entity);
+                // Paper end
 
                 guardEntityList = false; // Spigot
                 this.entityList.remove(this.tickPosition--); // CraftBukkit - Use field for loop variable
@@ -1324,7 +1332,7 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
                 BlockPosition blockposition = tileentity.getPosition();
 
                 // Paper start - Skip ticking in chunks scheduled for unload
-                net.minecraft.server.Chunk chunk = this.getChunkIfLoaded(blockposition);
+                net.minecraft.server.Chunk chunk = tileentity.getCurrentChunk();
                 boolean shouldTick = chunk != null;
                 if(this.paperConfig.skipEntityTickingInChunksScheduledForUnload)
                     shouldTick = shouldTick && !chunk.isUnloading() && chunk.scheduledForUnload == null;
@@ -1360,8 +1368,11 @@ public abstract class World implements GeneratorAccess, IIBlockAccess, AutoClose
                 tilesThisCycle--;
                 this.tileEntityListTick.remove(tileTickPosition--);
                 //this.tileEntityList.remove(tileentity); // Paper - remove unused list
-                if (this.isLoaded(tileentity.getPosition())) {
-                    this.getChunkAtWorldCoords(tileentity.getPosition()).d(tileentity.getPosition());
+                // Paper start
+                net.minecraft.server.Chunk chunk = tileentity.getCurrentChunk();
+                if (chunk != null) {
+                    chunk.removeTileEntity(tileentity.getPosition());
+                    // Paper end
                 }
             }
         }
-- 
2.18.0

